modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java

```diff
@@ -0,0 +1,272 @@
+    static final String DATABASES_INDEX = ".geoip_databases";
+    static final int MAX_CHUNK_SIZE = 1024 * 1024;
+
+    private final Client client;
+    private final HttpClient httpClient;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final Logger logger = LogManager.getLogger(GeoIpDownloader.class);
+    private final PersistentTasksService persistentTasksService;
+
+    //visible for testing
+    protected volatile GeoIpTaskState state;
+    private volatile TimeValue pollInterval;
+    private volatile String endpoint;
+    private volatile AllocatedPersistentTask persistentTask;
+    private volatile Scheduler.ScheduledCancellable scheduled;
+    private volatile boolean enabled;
+
+    GeoIpDownloader(Client client, HttpClient httpClient, ClusterService clusterService, ThreadPool threadPool, Settings settings) {
+        super(GEOIP_DOWNLOADER, ThreadPool.Names.GENERIC);
+        this.client = new OriginSettingClient(client, "geoip");
+        this.httpClient = httpClient;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        persistentTasksService = new PersistentTasksService(clusterService, threadPool, client);
+        endpoint = ENDPOINT_SETTING.get(settings);
+        pollInterval = POLL_INTERVAL_SETTING.get(settings);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(POLL_INTERVAL_SETTING, this::setPollInterval);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ENDPOINT_SETTING, this::setEndpoint);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+        enabled = ENABLED_SETTING.get(settings);
+        if (enabled) {
+            clusterService.addListener(this);
+        }
+    }
+
+    public void setPollInterval(TimeValue pollInterval) {
+        this.pollInterval = pollInterval;
+        if (scheduled != null && scheduled.cancel()) {
+            scheduleNextRun(new TimeValue(1));
+        }
+    }
+
+    public void setEndpoint(String endpoint) {
+        this.endpoint = endpoint;
+    }
+
+    public void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        if (enabled) {
+            startPersistentTask();
+        } else {
+            if (scheduled != null) {
+                scheduled.cancel();
+            }
+            persistentTasksService.sendRemoveRequest(GEOIP_DOWNLOADER, ActionListener.wrap(r -> {
+                state = ((GeoIpTaskState) r.getState()).copy();
+            }, e -> logger.error("could not remove task [" + GEOIP_DOWNLOADER + "]", e)));
+        }
+    }
+
+    void updateDatabases() throws IOException {
+        logger.info("updating geoip databases");
+        String data = httpClient.getString(endpoint + "?key=11111111-1111-1111-1111-111111111111");
+        List<Map<String, Object>> response = XContentHelper.convertToList(XContentType.JSON.xContent(), data, true);
```

Comment:
maybe replace this logic with something like a helper method and invoke that from this method?

```
@SuppressWarnings("unchecked")
    <T> List<T> fetchDatabasesOverview() throws IOException {
        byte[] data = httpClient.getBytes(endpoint + "?key=11111111-1111-1111-1111-111111111111");
        try (XContentParser parser = XContentType.JSON.xContent().createParser(NamedXContentRegistry.EMPTY,
            DeprecationHandler.THROW_UNSUPPORTED_OPERATION, data)) {
            return (List<T>) parser.list();
        }
    }
```

This way no changes are needed to `XContentHelper` and `CompressFactory`?

Reply:
Good idea, done